### DISCUSSION TITLE
[SDK] Error while setting custom paymaster data with Go-SDK

### DISCUSSION THREAD

**Discussion by hoshiyari420** [zkSync-Community-Hub/zksync-developers#811] [SDKs] [answered]
### Team or Project

Zyfi

### Select the SDK

Go

### Environment

Testnet

### Issue Description

Error while trying to integrate Zyfi's paymaster with go-sdk. Returns the result with "failed to call eth_sendRawTransaction: failed to validate the transaction. reason: Validation revert: Paymaster validation error: Error function_selector = 0x, data = 0x" 

I suspect that paymasterInput is not being sent correctly through sdk.

While the same transaction with same paymasterInput works fine with typescript. 

### Expected Behavior

Transaction should be successful with paymaster integration.

### Code Example

Here is the test code : 
go test -timeout 30s -run ^TestSendTransactionWithPaymaster$ paymaster
```
package main

import (
	"context"
	"fmt"
	"math/big"
	"strings"
	"testing"

	"github.com/ethereum/go-ethereum/common"
	"github.com/stretchr/testify/assert"
	"github.com/zksync-sdk/zksync2-go/accounts"
	"github.com/zksync-sdk/zksync2-go/clients"
	zkTypes "github.com/zksync-sdk/zksync2-go/types"
)

func TestSendTransactionWithPaymaster(t *testing.T) {
	zkSyncEraProvider := "https://sepolia.era.zksync.dev"
	rewardsContractAddress := "0xba8a8ff7E7332f7e05205Ec9fC927965435C552c"
	data := "0x0cd9181400000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000160000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000019000000000000000000000000000000000000000000000000000000000000000200000000000000000000000045005ac8be07a914ba9b3a1530f50eab2b297b5f00000000000000000000000099966fbddc86252b24ea2429de0b80631085b2dc0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000006400000000000000000000000000000000000000000000000000000000000000c800000000000000000000000000000000000000000000000000000000000000411801d27688304159418d744a582c8189441effb3537e6defa9cd0494607848387305d5f4db4b42d22f88722886f9c7289b9d931b3def10d9e3748ffa13f837fc1c00000000000000000000000000000000000000000000000000000000000000"
	gasLimit := "15000000"
	maxFeePerGas := "25000000"
	gasPerPubdata := "50000"
	paymaster := "0xC530313f1AF0B2B3A41DB2C20D88aF1c93A4f878"

	// Paymaster input will be expired in 1 hour, so you need to make a new requets to get a new one with your sender address : 0x497e3aD1B769bC6F46E90f2b87692a2F3F40cf40
	paymasterInput := "0x949431dc000000000000000000000000fd1fbff2e1baa053c927dc513579a8b2727233d80000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000001200000000000000000000000000000000000000000000000000000019391e99b0e00000000000000000000000000000000000000000000000000000000000000a6000000000000000000000000497e3ad1b769bc6f46e90f2b87692a2f3f40cf40000000000000000000000000000000000000000000000000000000000000271000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000416b1fc48ef3da009a09da88be1f0cf33302ffdb3c90294648afcb705607e4f1414b1f042ac328b1e64c0d1ff9d7277355a413558336ae4c532d3ce2182a177e571c00000000000000000000000000000000000000000000000000000000000000"

	// Sender private key must consist of data and paymaster input
	senderPrivateKey := ""

	zkClient, err := clients.Dial(zkSyncEraProvider)
	assert.NoError(t, err)

	wallet, err := accounts.NewWallet(common.Hex2Bytes(senderPrivateKey), zkClient, nil)
	assert.NoError(t, err)
	fmt.Println((wallet.Address()));
	paymasterParams := &zkTypes.PaymasterParams{
		Paymaster:      common.HexToAddress(paymaster),
		PaymasterInput: common.Hex2Bytes(strings.Trim(paymasterInput, "0x")),
	}

	calldata := common.Hex2Bytes(strings.Trim(data, "0x"))

	gasFeeCapBN, ok := big.NewInt(0).SetString(maxFeePerGas, 10)
	assert.True(t, ok)

	gasLimitBN, ok := big.NewInt(0).SetString(gasLimit, 10)
	assert.True(t, ok)

	gasPerPubdataBN, ok := big.NewInt(0).SetString(gasPerPubdata, 10)
	assert.True(t, ok)

	contractAddr := common.HexToAddress(rewardsContractAddress)
	fmt.Println(paymasterParams)
	hash, err := wallet.SendTransaction(context.Background(), &accounts.Transaction{
		ChainID:         wallet.Signer().ChainID(),
		To:              &contractAddr,
		Data:            calldata,
		PaymasterParams: paymasterParams,
		Gas:             gasLimitBN.Uint64(),
		GasFeeCap:       gasFeeCapBN,
		GasPerPubdata:   gasPerPubdataBN,
	})
	fmt.Println(err)
	assert.NoError(t, err)

	// Wait for transaction to be finalized
	_, err = zkClient.WaitMined(context.Background(), hash)
	assert.NoError(t, err)

	fmt.Println("Transaction hash: ", hash.Hex())
}

```

### Repo Link (Optional)

https://github.com/trsmarc/zyfi-paymaster-go

**Comment by alhenaai** [comment]
It appears that the issue you're encountering with the Go SDK and Zyfi's paymaster integration might be related to how the `paymasterInput` is being handled or formatted in the Go SDK compared to the TypeScript implementation. The error message "failed to call eth_sendRawTransaction: failed to validate the transaction. reason: Validation revert: Paymaster validation error: Error function_selector = 0x, data = 0x" suggests that the paymaster validation is failing due to incorrect or missing data in the function selector or the data field.

Here are a few steps you can take to troubleshoot and potentially resolve this issue:

1. **Verify `paymasterInput` Format:** Ensure that the `paymasterInput` is correctly formatted and includes all necessary data. It's crucial that this input matches exactly what the paymaster expects, including the function selector and data.

2. **Compare SDK Implementations:** Since the transaction works with TypeScript, compare the exact format and data being sent through the TypeScript implementation with what is being sent through the Go SDK. There might be subtle differences in how data is encoded or handled.

3. **Check Paymaster Contract:** Review the paymaster contract's expected inputs and ensure that the Go SDK is sending data in the correct format and order. Pay special attention to how the function selector and data fields are constructed.

4. **Debugging and Logs:** Add more logging around the transaction creation and sending process in the Go SDK to capture what data is being sent exactly. This might help identify discrepancies or missing elements.

5. **Consult Documentation and Community:** Review the [Zyfi and ZKsync documentation](https://docs.zksync.io/) for any specifics about paymaster integration and check if there are updates or notices regarding SDK changes. You can also discuss this issue with the community or the developers through [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions) for more targeted advice.

If after these steps the issue persists, it might be beneficial to reach out directly to the Zyfi or ZKsync support or development team with details of what you have tried and the specifics of the error. They might be able to provide more direct insights or updates regarding SDK compatibility or known issues.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by bxpana** [comment]
Thanks for the post, will check this out. Are you able to add me to the repo too?

**Reply by hoshiyari420** [reply]
sure, @trsmarc has sent you an invitation

**Comment by dutterbutter** [comment]
@hoshiyari420 hello sir 👋 

I would suggest updating `paymasterInput` to make use of the `utils.GetPaymasterParams` to ensure the params are properly formed (e.g. with `types.ApprovalBasedPaymasterInput` or general). Here are some example usage:

- [`use-paymaster.go`  ](https://github.com/zksync-sdk/zksync2-examples/blob/main/go/18_use_paymaster.go)
- [`paymaster test case`](https://github.com/zksync-sdk/zksync2-go/blob/main/test/account_abstraction_test.go)

**Reply by hoshiyari420** [reply]
@dutterbutter Hi ser! 

Since we are getting the `paymasterInput` data back from Zyfi API, it is already formed with the required paymaster flow "ApprovalBased" and with other params required for Zyfi's paymaster. Changing it might lead to validation issues from Zyfi paymasters directly.

**Reply by dutterbutter** [reply]
Okay, a couple followup questions:

- How is the data being formed on the API? Can you show me how this is current being done or can you confirm this is the same paymaster input data being used for other language implementations (e.g. you mentioned TypeScript)
- Can you confirm the paymaster has sufficient funds?
- How can I reproduce on my side to further debug / assist?

**Reply by hoshiyari420** [reply]
@dutterbutter  happy to answer! 

- On the API side, data is formed using `zksync-ethers` utils i.e. `getPaymasterParams`. Yes, I and @trsmarc can confirm that the same paymaster input are working for Typescript. we have added a script in the mentioned repo.

- Yes, I can confirm the paymaster has sufficient funds and transactions are being successfully propagated. https://sepolia.explorer.zksync.io/address/0xC530313f1AF0B2B3A41DB2C20D88aF1c93A4f878

- @trsmarc please add @dutterbutter to the repo as well. Easiest way we can assist with testing is that we provide you with the required calldata and paymasterInput once you provide a testing address and you can check if on typescript script to confirm the provided datas are correct.

**Reply by dutterbutter** [reply]
Thank you sir 🙏

**Reply by trsmarc** [reply]
I've sent you an invitation for the repo, please check it out

**Reply by dutterbutter** [reply, marked as answer]
**Update:**

I’ve created a [min-example repo](https://github.com/dutterbutter/debug-paymaster-input) to assist with debugging against a running fork of Sepolia using [anvil-zksync](https://github.com/matter-labs/anvil-zksync). I added some console logs to get more details during the Paymaster's `validateAndPayForPaymasterTransaction` function, where the issue seems to occur. 

### Findings:
It seems that the Paymaster input differs between the Go test and the JS test when interacting with the contract. Specifically, during the Go test, the Paymaster contract fails while decoding this line:

```solidity
(address token, uint256 amount, bytes memory data) = abi.decode(
    _transaction.paymasterInput[4:],
    (address, uint256, bytes)
);
```

### Observed Differences:
`_transaction.paymasterInput[4:]` differs as shown below:

**JS Test Input:**
```
000000000000000000000000fd1fbff2e1baa053c927dc513579a8b2727233d8000000000000000000000000000000000000000000000000000000000000cd140000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000012000000000000000000000000000000000000000000000000000000193ae6e16a80000000000000000000000000000000000000000000000000000000000000297000000000000000000000000497e3ad1b769bc6f46e90f2b87692a2f3f40cf40000000000000000000000000000000000000000000000000000000000000138800000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000004113904e6ebd358bc8628b3ff86b21b23402ed7f4b0a8ba7fc7efdfe60bd973c0d5bc084d19ecc845445274ce8314994f5fc7d8de98e9dd43ea7d6d4d1310c993e1c00000000000000000000000000000000000000000000000000000000000000
```

**Go Test Input:**
```
000000000000000000000000fd1fbff2e1baa053c927dc513579a8b2727233d8000000000000000000000000000000000000000000000000000000000000cd140000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000012000000000000000000000000000000000000000000000000000000193ae6e16a80000000000000000000000000000000000000000000000000000000000000297000000000000000000000000497e3ad1b769bc6f46e90f2b87692a2f3f40cf40000000000000000000000000000000000000000000000000000000000000138800000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000004113904e6ebd358bc8628b3ff86b21b23402ed7f4b0a8ba7fc7efdfe60bd973c0d5bc084d19ecc845445274ce8314994f5fc7d8de98e9dd43ea7d6d4d1310c993e1c
```

### Root Cause:
The issue lies in how `paymasterInput` is parsed in the Go test. Specifically:

```go
PaymasterInput: common.Hex2Bytes(strings.Trim(paymasterInput, "0x")),
```

### Fix:
Update the code: `strings.TrimPrefix` instead of `strings.Trim`:

**After Fix:**
```go
PaymasterInput: common.Hex2Bytes(strings.TrimPrefix(paymasterInput, "0x")),
```

Let me know if this small change resolves the issue, or if further debugging is required!

**Reply by hoshiyari420** [reply]
Hi @dutterbutter
Thank you so much for taking your time to look into this. Appreciate the example repo, it will help us debug such issues faster!
